pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java

```diff
@@ -1479,6 +1486,10 @@ public void deleteLedgerComplete(Object ctx) {
                         unfenceTopicToResume();
                     }
                 });
+
+            FutureUtil.completeAs(closeWithoutWaitingClientFuture, res);
+            FutureUtil.completeAs(fullyCloseFuture, res);
```

Comment:
What's the difference of `closeWithoutWaitingClientFuture` and `fullyCloseFuture`? Is it the same in this case?

Reply:
> What's the difference between `closeWithoutWaitingClientFuture` and `fullyCloseFuture`? Is it the same in this case?


We use `closeWithoutWaitingClientFuture` and `fullyCloseFuture` to prevent a new closing if there is an in-progress deletion or closing. We make Pulsar return the in-progress one when it is called the second time.

### Topic close 
- `PersistentTopic.delete(false /* closeWithoutWaitingClientDisconnect */)` will be executed directly; do not wait for the clients to be disconnected.
- `PersistentTopic.delete(true /* closeWithoutWaitingClientDisconnect */)` will be executed after the clients disconnected.

If the first one closing is called by `PersistentTopic.delete(true /* closeWithoutWaitingClientDisconnect */)`, the both `closeWithoutWaitingClientFuture` and `fullyCloseFuture` will be initialized as "waiting for clients disconnect" .

If the first one closing is called by `PersistentTopic.delete(false /* closeWithoutWaitingClientDisconnect */)`, `fullyCloseFuture` will be initialized as "waiting for clients disconnect"; and `closeWithoutWaitingClientFuture` will be initialized as "not waiting for clients disconnect" . 

### Topic delete 
- `PersistentTopic.delete(false /* closeIfClientsConnected */)` will fail when some clients are connected.
- `PersistentTopic.delete(true /* closeIfClientsConnected */)` will be executed after the clients disconnected.

the both of `closeWithoutWaitingClientFuture` and `fullyCloseFuture` will be initialized as "waiting for clients disconnect" .